### PR TITLE
Cleanup non-const methods in dense and combined features

### PR DESCRIPTION
**Pull request by vinx13**
* methods made const: `create_merged_copy`, `get_feature_matrix`, `get_feature_obj`
* dropped public methods: `set_feature_vector`, `reshape`, `steal_feature_matrix`
* made `free_features` protected

### CODE REVIEW

**Comment by lisitsyn** [comment]
lgtm!

File: src/shogun/classifier/NearestCentroid.cpp [hunk 1]
```diff
@@ -100,12 +92,10 @@ namespace shogun{
...
 		}
 
-		m_centroids->free_feature_matrix();
-		m_centroids->set_feature_matrix(centroids);
-
+		auto centroids_feats = some<CDenseFeatures<float64_t>>(centroids);
 
 		m_is_trained=true;
-		distance->init(m_centroids,distance->get_rhs());
+		distance->init(centroids_feats, distance->get_rhs());
```
Comment (karlnapf) [inline src/shogun/classifier/NearestCentroid.cpp]: why the change here?

File: src/shogun/features/CombinedFeatures.h [hunk 2]
```diff
@@ -76,7 +76,7 @@ class CCombinedFeatures : public CFeatures
 		}
 
 		/** list feature objects */
-		void list_feature_objs();
+		void list_feature_objs() const;
```
Comment (karlnapf) [inline src/shogun/features/CombinedFeatures.h]: lol, this method should rather be deleted (or return a list of strings or so)

File: src/shogun/features/DenseFeatures.cpp [hunk 3]
```diff
@@ -170,28 +170,6 @@ template<class ST> ST* CDenseFeatures<ST>::get_feature_vector(int32_t num, int32
 	return feat;
 }
 
-template<class ST> void CDenseFeatures<ST>::set_feature_vector(SGVector<ST> vector, int32_t num)
```
Comment (karlnapf) [inline src/shogun/features/DenseFeatures.cpp]: it is not used?

File: src/shogun/features/DenseFeatures.cpp [hunk 4]
```diff
@@ -331,16 +309,6 @@ void CDenseFeatures<ST>::copy_feature_matrix(SGMatrix<ST> target, index_t column
 	}
 }
 
-template<class ST> SGMatrix<ST> CDenseFeatures<ST>::steal_feature_matrix()
```
Comment (karlnapf) [inline src/shogun/features/DenseFeatures.cpp]: lol another joker method :D

File: src/shogun/features/DenseFeatures.cpp [hunk 5]
```diff
@@ -468,21 +438,6 @@ template<class ST> void CDenseFeatures<ST>::initialize_cache()
 
 template<class ST> EFeatureClass CDenseFeatures<ST>::get_feature_class() const  { return C_DENSE; }
 
-template<class ST> bool CDenseFeatures<ST>::reshape(int32_t p_num_features, int32_t p_num_vectors)
```
Comment (karlnapf) [inline src/shogun/features/DenseFeatures.cpp]: this should just be in the matrix itself, i.e. create a new feature instance using the reshaped sgmatrix

File: src/shogun/features/Features.h [hunk 6]
```diff
@@ -259,7 +259,7 @@ class CFeatures : public CSGObject
 		 * @return new feature object which contains copy of data of this
 		 * instance and of given one
 		 */
-		virtual CFeatures* create_merged_copy(CFeatures* other)
+		virtual CFeatures* create_merged_copy(CFeatures* other) const
```
Comment (karlnapf) [inline src/shogun/features/Features.h]: maybe we can rename this to `merge`  at some point later?

**Comment by karlnapf** [comment]
CI is not happy though

File: src/shogun/features/DenseFeatures.cpp (hunk 3 above)
Comment (vinx13) [reply src/shogun/features/DenseFeatures.cpp]: yes, and i think `set_feature_matrix` should be also removed (after some refactor)

File: src/shogun/classifier/NearestCentroid.cpp (hunk 1 above)
Comment (vinx13) [reply src/shogun/classifier/NearestCentroid.cpp]: since we would like to drop setters in features`set_num_vectors`, `set_num_features`, we can't an empty features and then set feature matrix later, and here there is no need to store it in a member field

File: src/shogun/features/DenseFeatures.cpp (hunk 3 above)
Comment (karlnapf) [reply src/shogun/features/DenseFeatures.cpp]: absolutely, the instance should be tied to the matrix, and otherwise users can create a new instance

**Comment by karlnapf** [comment]
let me know once CI is happy and I will merge

**Comment by vinx13** [comment]
@karlnapf timeout on windows

**Comment by karlnapf** [comment]
restarted.
But I think this is good. Merging
